show generic parameter sig help after conditional access

brettfo opened the pull request:
Fixes #1598

This is a crashing bug and is a candidate for preview.

Note that bug #1628 was found during testing, but since it's non-crashing that bugfix will not be taken in the preview branch.

brettfo commented:
Tagging @Pilchie @dpoeschl @jasonmalinowski @rchande @balajikris @basoundr as reviewers for a preview bugfix.

Pilchie commented:
:+1:

rchande commented on src/Workspaces/CSharp/Portable/Extensions/SimpleNameSyntaxExtensions.cs:
```diff
@@ -15,7 +15,15 @@ public static ExpressionSyntax GetLeftSideOfDot(this SimpleNameSyntax name)
+                else
+                {
+                    return ((MemberAccessExpressionSyntax)name.Parent).Expression;
+                }
```
Use SyntaxNodeExtensions.GetParentConditionalAccessExpression. It always finds the right node, even with more complicated conditional access.

Pilchie commented on src/Workspaces/CSharp/Portable/Extensions/SimpleNameSyntaxExtensions.cs:
```diff
@@ -15,7 +15,15 @@ public static ExpressionSyntax GetLeftSideOfDot(this SimpleNameSyntax name)
             if (name.IsMemberAccessExpressionName())
             {
-                return ((MemberAccessExpressionSyntax)name.Parent).Expression;
+                var conditionalAccess = name.Parent.Parent as ConditionalAccessExpressionSyntax;
```
Is this only an issue if the `ConditionalAccessExpressionSyntax` is the grandparent?  What if it's further up a chain of dotting?

rchande commented on src/EditorFeatures/Test2/IntelliSense/CSharpSignatureHelpCommandHandlerTests.vb:
```diff
@@ -356,5 +358,28 @@ class C
+    }
+}
+]]></Document>)
+
```
Tests for more complicated cases? :smile:

balajikris replied:
like nested conditional access expressions..

Pilchie commented:
Also, make sure to run the tests not covered by Jenkins, in case that extension is used by other features.

balajikris commented:
I know VB parses this slightly differently, but have you checked that it doesn't crash in VB too?

rchande commented:
In case you need it, VB has the (arbitrarily differently named) GetCorrespondingConditionalAccessExpression helper.

brettfo commented:
I've switched to use `SyntaxNodeExtensions.GetParentConditionalAccessExpression` as suggested, added more complicated C# tests, and I also added VB tests to ensure there's no regression (no fix for VB was needed.)

I've also run all open and closed tests on my box.

rchande commented:
:+1:

Pilchie commented:
:+1:
